### PR TITLE
fix(54465): Broken emit with private field in class decorator

### LINKED ISSUE #54465: Broken emit with private field in class decorator
# Bug Report

### 🔎 Search Terms

private field, decorators, broken emit

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about decorators

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?useDefineForClassFields=true&module=7&ts=5.1.0-beta&ssl=15&ssc=1&pln=2&pc=1#code/FAMwrgdgxgLglgewgAgLYEMDWBTAItqBAJ3RmIAoQIAuZcgOkfSIHMBnW9CATwG0BdAJTIAvAD5kXbsIDewZAuRFsMMERTho8JMgAmBYqWwBhADbo2bADzHk2AB4xsEXW2QRsAdzqN6zdpw8AsLikjxi5PKK0ZjmlrTGADTIUdEKsPYJcWz4hCRkRMZITo6JqcjCyHJpigD0tcgAKgAW2MhwqAAOptiozjCkiCgIIMgwzXBuugjYbBAA5DBopE5E5QC+wJvAAAIYOLmGBeT2ohL29ADEIAgIgsBQ2cgAYrdV5de3tBBgqABG2CIomQAGYANxbYBAA)

### 💻 Code

```ts
function makeDecorator(fn: (...args: any[]) => any) {
    return function decorateClass<C extends new (...args: any[]) => any>(
        klass: C, 
        ctx: ClassDecoratorContext,
    )  {
        // The implementation of this doesn't matter
    }
}

@makeDecorator(x => x.#foo)
class Foo {
    #foo: number = 3;
}

```

### 🙁 Actual behavior

The generated code is syntatically incorrect with the error:

```
Unexpected token ')' 
```

This happens on the line, basically nothing is emitted after the `x.`:

```
    let _classDecorators = [makeDecorator(x => x.)];

```

### 🙂 Expected behavior

Functioning code should be emitted.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #54465

> https://github.com/microsoft/TypeScript/issues/54465#issuecomment-1572709502

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -31515,6 +31515,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
             }
 
+            if (findAncestor(right.parent, n => isDecorator(n) && isClassLike(n.parent))) {
+                grammarErrorOnNode(right, Diagnostics.Private_identifiers_are_not_allowed_outside_class_bodies);
+                return anyType;
+            }
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:31521]: I think this incorrectly issues an error on the following code:

```ts
class C {
    #myProp = "hello!";
    method() {
        @decorator(() => this.#myProp)
        class C {}

        return C;
    }
}

new C().method();


function decorator(f: () => string): <T>(x: T, ctxt: any) => T {
    console.log(f());
    return (x, _ctxt) => x;
}
```

@rbuckton can confirm.

Comment (rbuckton) [reply src/compiler/checker.ts]: Yes, a class decorator should only have access to private identifiers defined in the same lexical scope but *outside* of the class it decorats.

Comment (a-tarasyuk) [reply src/compiler/checker.ts]: Classes offer tremendous flexibility 😄. What are your thoughts on using this approach to detect the scope of private fields? Perhaps there is a more efficient alternative/existing helper...

Comment (DanielRosenwasser) [reply src/compiler/checker.ts]: Hm - I kind of wish we could just do this check when a private field fails to resolve - but then it wouldn't be a grammar check.

Comment (DanielRosenwasser) [reply src/compiler/checker.ts]: Then again, we provide no error on the following JavaScript

```js
class C {
    method() { this.#wat }
}
```

Whereas most engines do error on this. So in a sense it's weird that we don't issue a *grammar* error on an invalid private name access.

That said, I'm not sure if we want to go there.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -10270,3 +10270,8 @@ export function getTextOfJsxNamespacedName(node: JsxNamespacedName) {
 export function intrinsicTagNameToString(node: Identifier | JsxNamespacedName) {
     return isIdentifier(node) ? idText(node) : getTextOfJsxNamespacedName(node);
 }
+
+/** @internal */
+export function getOutermostDecoratedClass(node: Node) {
+    return findAncestor(node.parent, n => isDecorator(n) && isClassLike(n.parent) && getContainingClass(n.parent.parent) === undefined);
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts:10276]: If you're concerned about walking to the top of the tree, you can always change this to short-circuit on the first class:

```suggestion
    return findAncestor(node.parent, n => isClassLike(n) ? "stop"
        : isDecorator(n) && isClassLike(n.parent) && getContainingClass(n.parent.parent) === undefined);
```

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -31515,6 +31516,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
             }
 
+            if (getOutermostDecoratedClass(right)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:31519]: I think the right name is `hasClassAncestorOutsideOfDecorator` or something like that.

Also, can you just use the new check inside of `checkGrammarPrivateIdentifierExpression`?

**Comment by rbuckton** [comment]
I think what we should be looking at is `lookupSymbolForPrivateIdentifierDeclaration`. Right now it just uses `getContainingClass(location)` which does not skip over the immediately containing class of a class decorator.

We need something more like:

```ts
function getContainingClassExcludingClassDecorators(node: Node) {
  const decorator = findAncestor(node.parent, n => isClassLike(n) ? "quit" : isDecorator(n));
  if (decorator && isClassLike(decorator.parent)) return getContainingClass(decorator.parent);
  return getContainingClass(decorator ?? node);
}
```

Which would then replace the first `getContainingClass` call in `lookupSymbolForPrivateIdentifierDeclaration`.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -31378,14 +31379,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     // Lookup the private identifier lexically.
-    function lookupSymbolForPrivateIdentifierDeclaration(propName: __String, location: Node): Symbol | undefined {
-        for (let containingClass = getContainingClass(location); !!containingClass; containingClass = getContainingClass(containingClass)) {
+    function lookupSymbolForPrivateIdentifierDeclaration(propName: __String, location: Node, containingClass = getContainingClassExcludingClassDecorators(location)): Symbol | undefined {
```
Comment (rbuckton) [inline src/compiler/checker.ts:31382]: I'm not sure I like passing in `containingClass` here as it results in `location` being unused.

Instead, we could make the third parameter something like `includeLocation?: boolean`, pass in the containing class you already resolved as `location`, and tweak the original `for` loop to use `location` if `includeLocation` is true and it's already a `ClassLike`.

Comment (rbuckton) [reply src/compiler/checker.ts:31382]: Either that or drop `location` and require callers always pass in `containingClass` themselves.

**Review by rbuckton — APPROVED** [review]

**Pull request merged**